packages/table/src/common/internal/focusedCellUtils.ts

```diff
@@ -26,13 +27,72 @@ export function getInitialFocusedCell(
-        // focus the top-left cell of the first selection
+        // focus the top-left cell of the last selection
+        const lastIndex = selectedRegions.length - 1;
         return {
-            ...Regions.getFocusCellCoordinatesFromRegion(selectedRegions[0]),
-            focusSelectionIndex: 0,
+            ...Regions.getFocusCellCoordinatesFromRegion(selectedRegions[lastIndex]),
+            focusSelectionIndex: lastIndex,
         };
     } else {
         // focus the top-left cell of the table
         return { col: 0, row: 0, focusSelectionIndex: 0 };
     }
 }
+
+/**
+ * Returns a new cell-coordinates object that includes a focusSelectionIndex property.
+ * The returned object will have the proper IFocusedCellCoordinates type.
+ */
+export function toFullCoordinates(
+    cellCoords: ICellCoordinates,
+    focusSelectionIndex: number = 0,
+): IFocusedCellCoordinates {
+    return { ...cellCoords, focusSelectionIndex };
+}
+
+/**
+ * Expands an existing region to new region based on the current focused cell.
+ * The focused cell is an invariant and should not move as a result of this
+ * operation. This function is used, for instance, to expand a selected region
+ * on shift+click.
+ */
+export function expandFocusedRegion(focusedCell: IFocusedCellCoordinates, newRegion: IRegion) {
+    switch (Regions.getRegionCardinality(newRegion)) {
+        case RegionCardinality.FULL_COLUMNS: {
+            const [indexStart, indexEnd] = getExpandedRegionIndices(focusedCell, newRegion, "col", "cols");
+            return Regions.column(indexStart, indexEnd);
+        }
+        case RegionCardinality.FULL_ROWS: {
+            const [indexStart, indexEnd] = getExpandedRegionIndices(focusedCell, newRegion, "row", "rows");
+            return Regions.row(indexStart, indexEnd);
+        }
+        case RegionCardinality.CELLS:
+            const [rowIndexStart, rowIndexEnd] = getExpandedRegionIndices(focusedCell, newRegion, "row", "rows");
+            const [colIndexStart, colIndexEnd] = getExpandedRegionIndices(focusedCell, newRegion, "col", "cols");
+            return Regions.cell(rowIndexStart, colIndexStart, rowIndexEnd, colIndexEnd);
+        default: // i.e. `case RegionCardinality.FULL_TABLE:`
+            return Regions.table();
+    }
+}
+
+function getExpandedRegionIndices(
+    focusedCell: IFocusedCellCoordinates,
+    newRegion: IRegion,
+    focusedCellDimension: "row" | "col",
+    regionDimension: "rows" | "cols",
+) {
+    const srcIndex = focusedCell[focusedCellDimension];
+    const [dstIndex, dstIndexEnd] = newRegion[regionDimension];
+
+    if (dstIndex !== dstIndexEnd) {
+        if (regionDimension === "rows") {
+            throw new Error(Errors.TABLE_EXPAND_FOCUSED_REGION_MULTI_ROW_REGION);
+        } else if (regionDimension === "cols") {
+            throw new Error(Errors.TABLE_EXPAND_FOCUSED_REGION_MULTI_COLUMN_REGION);
+        }
+    }
+
+    return srcIndex <= dstIndex
```

Comment:
Oh, yeah, 'cause you have src. Ok. I guess dst is probably fine? Still slight preference for spelling both of them out, though.

Reply:
Opted for the equal-length variable names for the aesthetic of it, but I'll spell them out.